File: bot/highlight_clipper/worker.py
from __future__ import annotations

import asyncio
import logging
import shutil
import time
from pathlib import Path

from .config import CLIPS_DIR
from .config import CLIP_POST_ROLL_SECONDS
from .config import CLIP_PRE_ROLL_SECONDS
from .config import MAX_CLIP_SECONDS
from .config import POLL_INTERVAL_SECONDS
from .deadlock_client import get_match_history
from .deadlock_client import get_match_metadata
from .demo_analyzer import KillMoment
from .demo_analyzer import analyze_match
from .demo_analyzer import detect_all_events
from .demo_analyzer import moments_to_events
from .demo_downloader import cleanup_demo
from .demo_downloader import get_demo_path
from .dm_sender import send_highlight_to_channel
from .event_detector import HighlightEvent
from .event_detector import detect_events
from .state import is_match_processed
from .state import load_state
from .state import mark_match_processed
from .twitch_vod import download_clip
from .twitch_vod import find_vod_for_match
from .twitch_vod import get_channel_id

log = logging.getLogger("TwitchStreams.HighlightClipper")

_STEAM64_BASE = 76561197960265728
_STEAM_LINKS_DB = Path("/home/naniadm/Documents/Deadlock-Bots/data/deadlock.sqlite3")

# Alle aktiven Partner mit Discord-User-ID aus der Twitch-Bot-DB
_PARTNERS_QUERY = """
    SELECT twitch_login, discord_user_id
      FROM twitch_streamers_partner_state
     WHERE is_partner_active = 1
       AND discord_user_id IS NOT NULL
     ORDER BY twitch_login
"""


class HighlightClipperWorker:
    def __init__(self, bot, *, cog=None) -> None:
        self.bot = bot
        self._cog = cog
        self._task: asyncio.Task[None] | None = None
        Path(CLIPS_DIR).mkdir(parents=True, exist_ok=True)

    async def start(self) -> None:
        if self._task is not None and not self._task.done():
            return
        self._task = asyncio.create_task(self._loop(), name="twitch.highlight_clipper")
        log.info("HighlightClipper worker started")

    async def stop(self) -> None:
        task = self._task
        self._task = None
        if task is None:
            return
        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass
        log.info("HighlightClipper worker stopped")

    async def _loop(self) -> None:
        await self.bot.wait_until_ready()
        while not self.bot.is_closed():
            try:
                await self._run_once()
            except asyncio.CancelledError:
                raise
            except Exception:
                log.exception("HighlightClipper: Fehler im Worker")
            await asyncio.sleep(POLL_INTERVAL_SECONDS)

    async def _run_once(self) -> None:
        twitch_api = self._get_twitch_api()
        if twitch_api is None:
            log.warning("HighlightClipper: TwitchAPI nicht verfügbar, überspringe")
            return

        streamers = await self._get_partner_streamers()
        if not streamers:
            log.info("HighlightClipper: Keine aktiven Partner mit Steam-ID gefunden")
            return
        log.info("HighlightClipper: %s Partner werden verarbeitet", len(streamers))

        state = load_state()
        now = int(time.time())

        for twitch_login, steam_id in streamers:
            try:
                await self._process_streamer(
                    state=state,
                    twitch_login=twitch_login,
                    steam_id=int(steam_id),
                    twitch_api=twitch_api,
                    now=now,
                )
            except asyncio.CancelledError:
                raise
            except Exception:
                log.exception("HighlightClipper: Fehler bei Streamer %s", twitch_login)

    async def _process_streamer(
        self,
        *,
        state: dict,
        twitch_login: str,
        steam_id: int,
        twitch_api,
        now: int,
    ) -> None:
        matches = await get_match_history(steam_id, limit=10)
        recent_matches = _filter_recent_matches(matches, state, login=twitch_login, now=now)
        if not recent_matches:
            return

        channel_id = await get_channel_id(twitch_login, twitch_api)
        if not channel_id:
            log.warning("HighlightClipper: Kein Twitch-Channel für %s", twitch_login)
            return

        for match in recent_matches:
            match_id = int(match["match_id"])
            clip_dir = (Path(CLIPS_DIR).resolve() / twitch_login / str(match_id))
            clip_dir.mkdir(parents=True, exist_ok=True)
            try:
                await self._process_match(
                    state=state,
                    twitch_login=twitch_login,
                    steam_id=steam_id,
                    match=match,
                    channel_id=channel_id,
                    twitch_api=twitch_api,
                    clip_dir=clip_dir,
                )
            finally:
                shutil.rmtree(clip_dir, ignore_errors=True)

    async def _process_match(
        self,
        *,
        state: dict,
        twitch_login: str,
        steam_id: int,
        match: dict,
        channel_id: str,
        twitch_api,
        clip_dir: Path,
    ) -> None:
        match_id = int(match["match_id"])
        match_start_unix = int(match["start_time"])
        match_duration_s = int(match.get("match_duration_s") or 0)

        # Demo-First: Events direkt aus Replay lesen, kein API-based detect_events mehr
        match_info = await get_match_metadata(match_id)
        hero_id = _get_hero_id(steam_id, match_info)
        events: list[HighlightEvent] = []

        demo_path = await get_demo_path(match_id)
        if demo_path is not None:
            try:
                moments = await detect_all_events(demo_path, hero_id or 0, twitch_login)
                log.info(
                    "HighlightClipper: Demo analysiert für %s match=%s — %s Kills (%s clutch)",
                    twitch_login, match_id, len(moments),
                    sum(1 for m in moments if m.is_clutch),
                )
                events = moments_to_events(moments, min_score=2)
            finally:
                cleanup_demo(match_id)

        # Fallback auf API-Erkennung wenn Demo nicht verfügbar
        if not events:
            api_events = detect_events(steam_id, match_info)
            if api_events:
                log.info("HighlightClipper: Demo-Analyse fehlgeschlagen, nutze API-Fallback für %s", twitch_login)
                events = api_events

        vod = await find_vod_for_match(
            channel_id,
            match_start_unix,
            match_duration_s,
            twitch_api,
        )
        if vod is None:
            log.warning(
                "HighlightClipper: Kein VOD für %s match_id=%s",
                twitch_login,
                match_id,
            )
            mark_match_processed(state, twitch_login, match_id)
            return

        clip_paths: list[str] = []
        clip_events: list[HighlightEvent] = []
        vod_offset_s = match_start_unix - int(vod["vod_started_at"])

        for index, event in enumerate(events, start=1):
            clip_start_s = max(0, vod_offset_s + event.game_time_s - CLIP_PRE_ROLL_SECONDS)
            clip_end_s = max(
                clip_start_s + 1,
                vod_offset_s + event.game_time_s + event.duration_s + CLIP_POST_ROLL_SECONDS,
            )
            clip_end_s = min(clip_end_s, clip_start_s + MAX_CLIP_SECONDS)
            output_path = clip_dir / f"{index:02d}_{event.event_type}_{event.game_time_s}.mp4"
            downloaded = await download_clip(
                str(vod["vod_id"]),
                clip_start_s,
                clip_end_s,
                str(output_path),
            )
            if not downloaded:
                continue
            clip_paths.append(str(output_path))
            clip_events.append(event)

        if clip_paths:
            await send_highlight_to_channel(
                self.bot,
                twitch_login,
                match_id,
                clip_events,
                clip_paths,
            )
        else:
            log.warning(
                "HighlightClipper: Keine Clips erstellt für %s match_id=%s",
                twitch_login,
                match_id,
            )

        mark_match_processed(state, twitch_login, match_id)

    def _get_twitch_api(self):
        if self._cog is not None:
            return getattr(self._cog, "api", None)
        cog = getattr(self.bot, "cogs", {}).get("TwitchStreamCog")
        return getattr(cog, "api", None) if cog is not None else None

    async def _get_partner_streamers(self) -> list[tuple[str, str]]:
        loop = asyncio.get_event_loop()
        try:
            rows = await loop.run_in_executor(None, _query_partner_streamers)
        except Exception:
            log.exception("HighlightClipper: DB-Abfrage für Partner fehlgeschlagen")
            rows = []

        # Discord user_id → account_id aus Steam-Bot SQLite auflösen
        discord_to_account: dict[int, str] = {}
        try:
            discord_ids = [int(row[1]) for row in (rows or []) if row[1]]
            if discord_ids:
                discord_to_account = await loop.run_in_executor(
                    None, _load_steam_account_ids, discord_ids
                )
        except Exception:
            log.exception("HighlightClipper: Steam-Links-Abfrage fehlgeschlagen")

        result: dict[str, str] = {}
        for row in (rows or []):
            login = str(row[0] or "").strip()
            discord_id = _as_int(row[1])
            if not login or discord_id is None:
                continue
            account_id = discord_to_account.get(discord_id)
            if account_id:
                result[login] = account_id

        result.update(_load_manual_steamids())
        return list(result.items())


def _load_manual_steamids() -> dict[str, str]:
    """Lädt manuelle Steam-ID-Zuordnungen aus data/highlight_clipper/steamids.json."""
    import json
    path = Path("data/highlight_clipper/steamids.json")
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(data, dict):
            return {str(k): str(v) for k, v in data.items() if k and v}
    except Exception:
        log.warning("HighlightClipper: steamids.json konnte nicht gelesen werden")
    return {}


def _query_partner_streamers() -> list:
    from ..storage.pg import query_all
    return query_all(_PARTNERS_QUERY) or []


def _load_steam_account_ids(discord_ids: list[int]) -> dict[int, str]:
    """Liest primary Steam-Account-IDs aus der Steam-Bot-SQLite für gegebene Discord-User-IDs."""
    import sqlite3
    if not _STEAM_LINKS_DB.exists():
        log.warning("HighlightClipper: Steam-Links-DB nicht gefunden: %s", _STEAM_LINKS_DB)
        return {}
    placeholders = ",".join("?" * len(discord_ids))
    conn = sqlite3.connect(f"file:{_STEAM_LINKS_DB}?mode=ro", uri=True)
    try:
        rows = conn.execute(
            f"SELECT user_id, steam_id FROM steam_links"
            f" WHERE user_id IN ({placeholders}) AND primary_account = 1",
            discord_ids,
        ).fetchall()
    finally:
        conn.close()
    result: dict[int, str] = {}
    for user_id, steam_id in rows:
        try:
            account_id = int(steam_id) - _STEAM64_BASE
            if account_id > 0:
                result[int(user_id)] = str(account_id)
        except (ValueError, TypeError):
            pass
    return result


def _filter_recent_matches(
    matches: list[dict],
    state: dict,
    *,
    login: str,
    now: int,
) -> list[dict]:
    min_start = now - 86400
    filtered: list[dict] = []
    for match in matches:
        if not isinstance(match, dict):
            continue
        match_id = _as_int(match.get("match_id"))
        start_time = _as_int(match.get("start_time"))
        if match_id is None or start_time is None:
            continue
        if start_time <= min_start or is_match_processed(state, login, match_id):
            continue
        filtered.append(
            {
                "match_id": match_id,
                "start_time": start_time,
                "match_duration_s": _as_int(match.get("match_duration_s")) or 0,
            }
        )
    filtered.sort(key=lambda item: item["start_time"])
    return filtered


def _as_int(value: object) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _get_hero_id(steam_id: int, match_info: dict) -> int | None:
    """Sucht die hero_id des Spielers anhand seiner account_id."""
    for player in (match_info.get("players") or []):
        if _as_int(player.get("account_id")) == steam_id:
            return _as_int(player.get("hero_id"))
    return None


def _score_events_with_demo(
    events: list[HighlightEvent],
    moments: list[KillMoment],
) -> list[HighlightEvent]:
    """
    Reichert Events mit Demo-Combo-Labels an und filtert Solo-Kills heraus.
    Events ohne passenden Demo-Moment bleiben unverändert.
    """
    from .event_detector import HighlightEvent as HE
    import dataclasses

    result = []
    for event in events:
        # Passendes Demo-Moment suchen (±5s Toleranz)
        best: KillMoment | None = None
        for m in moments:
            if abs(m.game_time_s - event.game_time_s) <= 5:
                if best is None or m.combo_score > best.combo_score:
                    best = m

        if best is None:
            # Kein Demo-Moment → Event unverändert behalten
            result.append(event)
            continue

        # Health-Label zusammenbauen
        hp_pct = round(best.health_pct * 100)
        health_label = ""
        if best.health_pct < 0.20:
            health_label = f"🔴 {hp_pct}% HP"
        elif best.health_pct < 0.35:
            health_label = f"🟠 {hp_pct}% HP"
        elif best.health_pct < 0.50:
            health_label = f"🟡 {hp_pct}% HP"

        parts = [event.label]
        if health_label:
            parts.append(health_label)
        if best.combo_label and best.combo_label != "Kill":
            parts.append(best.combo_label)
        label = " — ".join(parts)

        # Clips nur wenn excitement_score > 0 ODER Teamfight/Close Fight
        if event.event_type in ("teamfight", "close_fight") or best.excitement_score > 0:
            result.append(dataclasses.replace(event, label=label))
        else:
            log.info(
                "HighlightClipper: Event @ %ss herausgefiltert (HP=%s%% score=%s)",
                event.game_time_s, hp_pct, best.excitement_score,
            )

    # Hochscorende Demo-Momente die kein API-Event abdeckt → eigenes Clip-Event
    covered_times = {e.game_time_s for e in events}
    for m in moments:
        if m.excitement_score < 3:
            continue
        if any(abs(m.game_time_s - t) <= 10 for t in covered_times):
            continue  # Bereits abgedeckt
        pre_roll = min(int(m.combo_score) * 3 + 15, 30)
        result.append(HighlightEvent(
            event_type="close_fight",
            game_time_s=int(m.game_time_s),
            duration_s=0,
            kill_count=1,
            label=f"Clutch Kill ({int(m.health_pct*100)}% HP) — {m.combo_label}",
            pre_roll_s=pre_roll,
        ))
        covered_times.add(m.game_time_s)

    return result or events  # Fallback: alle Events wenn Demo-Filtering nichts übrig lässt

File: rust/crates/tb-highlight/src/worker.rs
//! Orchestrierung des Highlight-Clippers (Poll-Loop über aktive Partner).
//!
//! Port von `bot/highlight_clipper/worker.py`. Aufgebaut in Teil-Slices:
//! - 9a (hier): reine Entscheidungs-Helfer ([`filter_recent_matches`],
//!   [`get_hero_id`], [`compute_clip_window`]) — ohne I/O, voll testbar.
//! - 9b/9c (folgt): Partner-Datenschicht (Postgres + Steam-SQLite + manuelle
//!   steamids.json) und der eigentliche Poll-Loop inkl. Twitch-API.
//!
//! Das ungenutzte `_score_events_with_demo` (kein Caller in Python) wird nicht
//! portiert.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use sqlx::PgPool;

use crate::config::{
    CLIPS_DIR, CLIP_POST_ROLL_SECONDS, CLIP_PRE_ROLL_SECONDS, MAX_CLIP_SECONDS, STATE_PATH,
};
use crate::event_detector::{detect_events, HighlightEvent};
use crate::state::{is_match_processed, mark_match_processed, HighlightState};
use crate::twitch_vod::TwitchVodApi;
use crate::{deadlock_client, demo_analyzer, demo_downloader, highlight_sender, partners, twitch_vod};

/// Ein zu verarbeitendes Match (gefiltert + normalisiert aus der Match-History).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RecentMatch {
    pub match_id: i64,
    pub start_time: i64,
    pub match_duration_s: i64,
}

/// Filtert die Match-History auf verarbeitbare Matches: Objekt-Form, gültige
/// `match_id`/`start_time`, jünger als 24h und noch nicht verarbeitet. Sortiert
/// aufsteigend nach `start_time` (Python `_filter_recent_matches`).
pub fn filter_recent_matches(
    matches: &[serde_json::Value],
    state: &HighlightState,
    login: &str,
    now: i64,
) -> Vec<RecentMatch> {
    let min_start = now - 86400;
    let mut filtered: Vec<RecentMatch> = Vec::new();
    for m in matches {
        let Some(obj) = m.as_object() else { continue };
        let (Some(match_id), Some(start_time)) = (
            as_int(obj.get("match_id")),
            as_int(obj.get("start_time")),
        ) else {
            continue;
        };
        if start_time <= min_start || is_match_processed(state, login, match_id) {
            continue;
        }
        filtered.push(RecentMatch {
            match_id,
            start_time,
            match_duration_s: as_int(obj.get("match_duration_s")).unwrap_or(0),
        });
    }
    filtered.sort_by_key(|m| m.start_time);
    filtered
}

/// Sucht die `hero_id` des Spielers (per `account_id == steam_id`) in den
/// Match-Metadaten (Python `_get_hero_id`).
pub fn get_hero_id(steam_id: i64, match_info: &serde_json::Value) -> Option<i64> {
    let players = match_info
        .get("players")
        .and_then(serde_json::Value::as_array)?;
    for player in players {
        if as_int(player.get("account_id")) == Some(steam_id) {
            return as_int(player.get("hero_id"));
        }
    }
    None
}

/// Berechnet das Clip-Fenster (start, end) im VOD aus dem VOD-Offset und dem
/// Event: Pre-Roll vor dem Event, Post-Roll danach, gedeckelt auf
/// `MAX_CLIP_SECONDS` (Python-Logik in `_process_match`).
pub fn compute_clip_window(vod_offset_s: i64, event: &HighlightEvent) -> (i64, i64) {
    let clip_start_s = (vod_offset_s + event.game_time_s - CLIP_PRE_ROLL_SECONDS).max(0);
    let clip_end_s = (clip_start_s + 1)
        .max(vod_offset_s + event.game_time_s + event.duration_s + CLIP_POST_ROLL_SECONDS);
    let clip_end_s = clip_end_s.min(clip_start_s + MAX_CLIP_SECONDS);
    (clip_start_s, clip_end_s)
}

/// `int(value)`-Semantik aus Python (`_as_int`): Zahlen gegen 0 gekürzt, Strings
/// nur als reine Ganzzahl, Bools → 0/1, sonst None.
fn as_int(value: Option<&serde_json::Value>) -> Option<i64> {
    match value {
        Some(serde_json::Value::Number(n)) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)),
        Some(serde_json::Value::String(s)) => s.trim().parse::<i64>().ok(),
        Some(serde_json::Value::Bool(b)) => Some(i64::from(*b)),
        _ => None,
    }
}

/// Laufzeit-Konfiguration des Highlight-Clippers (Pfade + Endpoints, alle
/// injizierbar). [`Self::new`] füllt alles außer den extern aufzulösenden
/// `boon`- und `yt-dlp`-Binärpfaden aus den Modul-Konstanten.
#[derive(Debug, Clone)]
pub struct HighlightClipperConfig {
    pub boon_path: PathBuf,
    pub yt_dlp_path: PathBuf,
    pub steamids_json_path: PathBuf,
    pub clips_dir: PathBuf,
    pub demo_cache_dir: PathBuf,
    pub state_path: PathBuf,
    pub highlight_api_url: String,
    pub deadlock_api_base: String,
}

impl HighlightClipperConfig {
    /// Baut die Konfiguration mit Default-Pfaden; nur `boon`/`yt-dlp` müssen
    /// (vom Aufrufer aufgelöst) übergeben werden.
    pub fn new(boon_path: PathBuf, yt_dlp_path: PathBuf) -> Self {
        Self {
            boon_path,
            yt_dlp_path,
            steamids_json_path: PathBuf::from(partners::STEAMIDS_JSON_DEFAULT),
            clips_dir: PathBuf::from(CLIPS_DIR),
            demo_cache_dir: PathBuf::from(demo_downloader::DEMO_CACHE_DIR),
            state_path: PathBuf::from(STATE_PATH),
            highlight_api_url: highlight_sender::HIGHLIGHT_API_URL.to_string(),
            deadlock_api_base: deadlock_client::DEADLOCK_API_BASE.to_string(),
        }
    }
}

/// Der Highlight-Clipper-Worker: hält Postgres-Pool, Twitch-API-Implementierung
/// und Konfiguration. Ein Durchlauf = [`Self::run_once`] (Python `_run_once`);
/// der Poll-Loop-Lifecycle wird beim Wiring in tb-bot aufgesetzt.
pub struct HighlightClipperWorker {
    pool: PgPool,
    api: Arc<dyn TwitchVodApi>,
    config: HighlightClipperConfig,
}

impl HighlightClipperWorker {
    pub fn new(pool: PgPool, api: Arc<dyn TwitchVodApi>, config: HighlightClipperConfig) -> Self {
        Self { pool, api, config }
    }

    /// Ein vollständiger Durchlauf über alle aktiven Partner.
    pub async fn run_once(&self) {
        let streamers = partners::get_partner_streamers(
            &self.pool,
            &self.config.steamids_json_path,
        )
        .await;
        if streamers.is_empty() {
            tracing::info!("HighlightClipper: Keine aktiven Partner mit Steam-ID gefunden");
            return;
        }
        tracing::info!(count = streamers.len(), "HighlightClipper: Partner werden verarbeitet");

        let mut state = crate::state::load_state(&self.config.state_path);
        let now = chrono::Utc::now().timestamp();

        for (login, account_id_str) in streamers {
            let Ok(account_id) = account_id_str.parse::<i64>() else {
                continue;
            };
            self.process_streamer(&mut state, &login, account_id, now).await;
        }
    }

    async fn process_streamer(
        &self,
        state: &mut HighlightState,
        login: &str,
        account_id: i64,
        now: i64,
    ) {
        let matches =
            deadlock_client::get_match_history(&self.config.deadlock_api_base, account_id, 10)
                .await
                .unwrap_or_default();
        let recent = filter_recent_matches(&matches, state, login, now);
        if recent.is_empty() {
            return;
        }

        let channel_id = match twitch_vod::get_channel_id(self.api.as_ref(), login).await {
            Some(c) => c,
            None => {
                tracing::warn!(login, "HighlightClipper: Kein Twitch-Channel");
                return;
            }
        };

        for m in &recent {
            let clip_dir = self
                .config
                .clips_dir
                .join(login)
                .join(m.match_id.to_string());
            if std::fs::create_dir_all(&clip_dir).is_err() {
                continue;
            }
            self.process_match(state, login, account_id, m, &channel_id, &clip_dir)
                .await;
            if let Err(error) = std::fs::remove_dir_all(&clip_dir) {
                tracing::warn!(
                    %error,
                    path = %clip_dir.display(),
                    "HighlightClipper: temp Clip-Verzeichnis konnte nicht geloescht werden"
                );
            }
        }
    }

    async fn process_match(
        &self,
        state: &mut HighlightState,
        login: &str,
        account_id: i64,
        m: &RecentMatch,
        channel_id: &str,
        clip_dir: &Path,
    ) {
        let match_id = m.match_id;
        let match_info = match deadlock_client::get_match_metadata(
            &self.config.deadlock_api_base,
            match_id,
        )
        .await
        {
            Ok(match_info) => match_info,
            Err(error) => {
                tracing::warn!(
                    %error,
                    login,
                    match_id,
                    "HighlightClipper: Match-Metadaten nicht ladbar"
                );
                serde_json::json!({})
            }
        };
        let hero_id = get_hero_id(account_id, &match_info);
        let mut events: Vec<HighlightEvent> = Vec::new();

        // Demo-First: Events direkt aus dem Replay lesen.
        if let Some(demo_path) = demo_downloader::get_demo_path(
            &self.config.deadlock_api_base,
            &self.config.demo_cache_dir,
            match_id,
        )
        .await
        {
            let moments = demo_analyzer::detect_all_events(
                &self.config.boon_path,
                &demo_path,
                hero_id.unwrap_or(0),
                login,
            )
            .await;
            let clutch = moments.iter().filter(|mm| mm.is_clutch).count();
            tracing::info!(
                login,
                match_id,
                kills = moments.len(),
                clutch,
                "HighlightClipper: Demo analysiert"
            );
            events = demo_analyzer::moments_to_events(&moments, 2);
            demo_downloader::cleanup_demo(&self.config.demo_cache_dir, match_id);
        }

        // Fallback auf API-Erkennung, wenn die Demo nichts liefert.
        if events.is_empty() {
            let api_events = detect_events(account_id, &match_info);
            if !api_events.is_empty() {
                tracing::info!(login, "HighlightClipper: Demo-Analyse leer, nutze API-Fallback");
                events = api_events;
            }
        }

        let vod = match twitch_vod::find_vod_for_match(
            self.api.as_ref(),
            channel_id,
            m.start_time,
            m.match_duration_s,
        )
        .await
        {
            Some(v) => v,
            None => {
                tracing::warn!(login, match_id, "HighlightClipper: Kein VOD gefunden");
                if let Err(error) =
                    mark_match_processed(state, &self.config.state_path, login, match_id)
                {
                    tracing::warn!(
                        %error,
                        login,
                        match_id,
                        "HighlightClipper: Processed-Marker konnte nicht gespeichert werden"
                    );
                }
                return;
            }
        };

        let vod_offset_s = m.start_time - vod.vod_started_at;
        let mut clip_paths: Vec<String> = Vec::new();
        let mut clip_events: Vec<HighlightEvent> = Vec::new();

        for (idx, event) in events.iter().enumerate() {
            let (clip_start, clip_end) = compute_clip_window(vod_offset_s, event);
            let output_path = clip_dir.join(format!(
                "{:02}_{}_{}.mp4",
                idx + 1,
                event.event_type.as_str(),
                event.game_time_s
            ));
            let ok = twitch_vod::download_clip(
                &self.config.yt_dlp_path,
                &vod.vod_id,
                clip_start,
                clip_end,
                &output_path,
            )
            .await;
            if !ok {
                continue;
            }
            clip_paths.push(output_path.to_string_lossy().into_owned());
            clip_events.push(event.clone());
        }

        if clip_paths.is_empty() {
            tracing::warn!(login, match_id, "HighlightClipper: Keine Clips erstellt");
        } else {
            highlight_sender::send_highlight_to_channel(
                &self.config.highlight_api_url,
                login,
                match_id,
                &clip_events,
                &clip_paths,
            )
            .await;
        }

        if let Err(error) = mark_match_processed(state, &self.config.state_path, login, match_id) {
            tracing::warn!(
                %error,
                login,
                match_id,
                "HighlightClipper: Processed-Marker konnte nicht gespeichert werden"
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::event_detector::EventType;
    use serde_json::json;

    fn ev(game_time_s: i64, duration_s: i64) -> HighlightEvent {
        HighlightEvent {
            event_type: EventType::Multikill,
            game_time_s,
            duration_s,
            kill_count: 2,
            label: "x".into(),
            pre_roll_s: 0,
        }
    }

    #[test]
    fn filter_recent_fenster_und_sortierung() {
        let now = 1_000_000;
        let state = HighlightState::new();
        let matches = vec![
            json!({"match_id": 1, "start_time": now - 100, "match_duration_s": 1800}), // frisch
            json!({"match_id": 2, "start_time": now - 90000}),                          // > 24h alt
            json!({"match_id": 3, "start_time": now - 50}),                             // frisch, neuer
            json!("kein-objekt"),                                                       // übersprungen
            json!({"start_time": now - 10}),                                            // ohne match_id
        ];
        let out = filter_recent_matches(&matches, &state, "nani", now);
        // 1 und 3 bleiben, sortiert nach start_time (1 vor 3).
        assert_eq!(out.len(), 2);
        assert_eq!(out[0].match_id, 1);
        assert_eq!(out[1].match_id, 3);
        assert_eq!(out[0].match_duration_s, 1800);
        assert_eq!(out[1].match_duration_s, 0); // Default ohne Feld
    }

    #[test]
    fn filter_recent_ueberspringt_verarbeitete() {
        let now = 1_000_000;
        let mut state = HighlightState::new();
        state.insert(
            "nani".into(),
            crate::state::StreamerState { processed_matches: vec![5], last_checked: 0 },
        );
        let matches = vec![json!({"match_id": 5, "start_time": now - 100})];
        assert!(filter_recent_matches(&matches, &state, "nani", now).is_empty());
    }

    #[test]
    fn hero_id_per_account() {
        let mi = json!({"players": [
            {"account_id": 111, "hero_id": 7},
            {"account_id": 222, "hero_id": 9},
        ]});
        assert_eq!(get_hero_id(222, &mi), Some(9));
        assert_eq!(get_hero_id(999, &mi), None);
        assert_eq!(get_hero_id(1, &json!({})), None);
    }

    #[test]
    fn clip_window_clamping() {
        // vod_offset 1000, event @ 50s, dur 5s. PRE=6, POST=4, MAX=40.
        // start = 1000+50-6 = 1044; end = max(1045, 1000+50+5+4=1059)=1059; cap 1044+40=1084 → 1059.
        assert_eq!(compute_clip_window(1000, &ev(50, 5)), (1044, 1059));
        // Lange Dauer wird auf MAX_CLIP gedeckelt.
        let (s, e) = compute_clip_window(0, &ev(100, 1000));
        assert_eq!(s, 94);
        assert_eq!(e, 94 + MAX_CLIP_SECONDS);
        // Negativer Start wird auf 0 geklemmt.
        let (s2, _) = compute_clip_window(0, &ev(0, 0));
        assert_eq!(s2, 0);
    }
}
